openstack/pkg/designate/client.go

```diff
@@ -0,0 +1,213 @@
+	createOpts := recordsets.CreateOpts{
+		Name:    record.Name + zone,
+		Type:    record.Type,
+		TTL:     int(record.TTL / time.Second),
+		Records: []string{record.Value},
+	}
+
+	exist, err := p.getRecordID(record.Name, zone)
+	if err != nil {
+		return fmt.Errorf("cannot get recordID: %v", err)
+	}
+
+	if exist != "" {
+		return errors.New("DNS record already exist")
+	}
+
+	_, err = recordsets.Create(p.dnsClient, p.zoneID, createOpts).Extract()
+	if err != nil {
+		return fmt.Errorf("cannot create DNS record: %v", err)
+	}
+
+	return nil
+}
+
+func (p *Provider) updateRecord(record libdns.Record, zone string) error {
+	updateOpts := recordsets.UpdateOpts{
+		TTL:     intToPointer(int(record.TTL / time.Second)),
+		Records: []string{record.Value},
+	}
+
+	// Update updates a recordset in a given zone
+	_, err := recordsets.Update(p.dnsClient, p.zoneID, p.recordID, updateOpts).Extract()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Provider) deleteRecord(record libdns.Record, zone string) error {
+	err := recordsets.Delete(p.dnsClient, p.zoneID, p.recordID).ExtractErr()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Provider) exportEnvVariables() error {
+	err := os.Setenv("OS_REGION_NAME", p.AuthOpenStack.RegionName)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_TENANT_ID", p.AuthOpenStack.TenantID)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_IDENTITY_API_VERSION", p.AuthOpenStack.IdentityApiVersion)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_PASSWORD", p.AuthOpenStack.Password)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_AUTH_URL", p.AuthOpenStack.AuthURL)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_USERNAME", p.AuthOpenStack.Username)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_TENANT_NAME", p.AuthOpenStack.TenantName)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	err = os.Setenv("OS_ENDPOINT_TYPE", p.AuthOpenStack.EndpointType)
+	if err != nil {
+		return fmt.Errorf("cannot set environment variable: %v", err)
+	}
+	return nil
+}
+
+func (p *Provider) auth(zoneName string) error {
+	err := p.exportEnvVariables()
+	if err != nil {
+		return err
+	}
+
+	opts, err := openstack.AuthOptionsFromEnv()
+	if err != nil {
+		return err
+	}
+
+	provider, err := openstack.AuthenticatedClient(opts)
+	if err != nil {
+		return err
+	}
+
+	dnsClient, err := openstack.NewDNSV2(provider, gophercloud.EndpointOpts{
+		Region: os.Getenv("OS_REGION_NAME"),
+	})
+	if err != nil {
+		return err
+	}
+	p.dnsClient = dnsClient
+
+	zoneID, err := p.setZoneID(zoneName)
```

Comment:
Why is the zone set on the provider? The zone is scoped for each function call, not the lifetime of the provider struct.

Reply:
Thanks, good point. I moved it to the other method.